File: mobile/lib/models/queue.dart
/// User-attention state. Independent of `Resolution`.
enum EntryState {
  new_('new'),
  seen('seen'),
  dismissed('dismissed');

  const EntryState(this.wireName);
  final String wireName;

  static EntryState fromJson(String value) {
    switch (value) {
      case 'new':
        return EntryState.new_;
      case 'seen':
        return EntryState.seen;
      case 'dismissed':
        return EntryState.dismissed;
      default:
        throw ArgumentError('Unknown entry state: $value');
    }
  }

  String toJson() => wireName;
}

/// Task-outcome state for `Category.request` entries only.
enum Resolution {
  unresolved('unresolved'),
  approved('approved'),
  denied('denied');

  const Resolution(this.wireName);
  final String wireName;

  static Resolution fromJson(String value) {
    switch (value) {
      case 'unresolved':
        return Resolution.unresolved;
      case 'approved':
        return Resolution.approved;
      case 'denied':
        return Resolution.denied;
      default:
        throw ArgumentError('Unknown resolution: $value');
    }
  }

  String toJson() => wireName;
}

const int maxEscalationLevel = 2;

class QueueEntry implements Comparable<QueueEntry> {
  final String eventId;
  final int tier;
  final int score;
  final EntryState state;
  final Resolution? resolution;
  final bool superseded;
  final int escalationLevel;
  final int seq;

  const QueueEntry({
    required this.eventId,
    required this.tier,
    required this.score,
    required this.state,
    this.resolution,
    this.superseded = false,
    this.escalationLevel = 0,
    required this.seq,
  });

  factory QueueEntry.fromJson(Map<String, dynamic> json) {
    return QueueEntry(
      eventId: json['event_id'] as String,
      tier: (json['tier'] as num).toInt(),
      score: (json['score'] as num).toInt(),
      state: EntryState.fromJson(json['state'] as String),
      resolution: json['resolution'] != null
          ? Resolution.fromJson(json['resolution'] as String)
          : null,
      superseded: json['superseded'] as bool? ?? false,
      escalationLevel: (json['escalation_level'] as num?)?.toInt() ?? 0,
      seq: (json['seq'] as num).toInt(),
    );
  }

  Map<String, dynamic> toJson() => {
    'event_id': eventId,
    'tier': tier,
    'score': score,
    'state': state.toJson(),
    if (resolution != null) 'resolution': resolution!.toJson(),
    'superseded': superseded,
    'escalation_level': escalationLevel,
    'seq': seq,
  };

  QueueEntry copyWith({
    String? eventId,
    int? tier,
    int? score,
    EntryState? state,
    Resolution? resolution,
    bool clearResolution = false,
    bool? superseded,
    int? escalationLevel,
    int? seq,
  }) {
    return QueueEntry(
      eventId: eventId ?? this.eventId,
      tier: tier ?? this.tier,
      score: score ?? this.score,
      state: state ?? this.state,
      resolution: clearResolution ? null : (resolution ?? this.resolution),
      superseded: superseded ?? this.superseded,
      escalationLevel: escalationLevel ?? this.escalationLevel,
      seq: seq ?? this.seq,
    );
  }

  bool get isLive => state != EntryState.dismissed && !superseded;

  /// Sort key: lower tier first, higher score first, newer seq first.
  @override
  int compareTo(QueueEntry other) {
    if (tier != other.tier) {
      return tier.compareTo(other.tier);
    }
    if (score != other.score) {
      return other.score.compareTo(score);
    }
    return other.seq.compareTo(seq);
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is QueueEntry &&
          runtimeType == other.runtimeType &&
          eventId == other.eventId &&
          tier == other.tier &&
          score == other.score &&
          state == other.state &&
          resolution == other.resolution &&
          superseded == other.superseded &&
          escalationLevel == other.escalationLevel &&
          seq == other.seq;

  @override
  int get hashCode =>
      eventId.hashCode ^
      tier.hashCode ^
      score.hashCode ^
      state.hashCode ^
      seq.hashCode;
}

File: core/agentdesk-model/src/queue.rs
//! Mutable per-event queue metadata. Events themselves are immutable; this
//! is where state, score and escalation live. See docs/DATA_MODEL.md and
//! docs/SYSTEM_DESIGN.md "Event lifecycle".

use serde::{Deserialize, Serialize};

use crate::event::{Category, EventId};

/// User-attention state. Independent of `Resolution`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EntryState {
    New,
    Seen,
    Dismissed,
}

/// Task-outcome state for `Category::Request` entries only. A dismissed
/// request that is still `Unresolved` is still blocking the agent.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Resolution {
    Unresolved,
    Approved,
    Denied,
}

/// Maximum escalation level; the watchdog never raises beyond this.
pub const MAX_ESCALATION_LEVEL: u8 = 2;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct QueueEntry {
    pub event_id: EventId,
    /// Derived from the event's category; fixed for the entry's lifetime.
    pub tier: u8,
    /// Recomputed on tick; orders within a tier. 0..=100.
    pub score: u16,
    pub state: EntryState,
    /// `Some` only for request entries.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub resolution: Option<Resolution>,
    /// Working entries of a task that has since closed.
    #[serde(default)]
    pub superseded: bool,
    /// Working entries only; 0..=MAX_ESCALATION_LEVEL.
    #[serde(default)]
    pub escalation_level: u8,
    /// Copied from the event for tie-breaking.
    pub seq: u64,
}

impl QueueEntry {
    /// A fresh entry for a newly processed event.
    pub fn new(event_id: EventId, category: Category, seq: u64, score: u16) -> Self {
        QueueEntry {
            event_id,
            tier: category.tier(),
            score,
            state: EntryState::New,
            resolution: (category == Category::Request).then_some(Resolution::Unresolved),
            superseded: false,
            escalation_level: 0,
            seq,
        }
    }

    /// Included in snapshots and re-scoring.
    pub fn is_live(&self) -> bool {
        self.state != EntryState::Dismissed && !self.superseded
    }

    /// Sort key: lower tier first, higher score first, newer first.
    pub fn order_key(&self) -> (u8, std::cmp::Reverse<u16>, std::cmp::Reverse<u64>) {
        (
            self.tier,
            std::cmp::Reverse(self.score),
            std::cmp::Reverse(self.seq),
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;
    use uuid::Uuid;

    #[test]
    fn new_request_entry_is_unresolved_and_others_have_no_resolution() {
        let r = QueueEntry::new(Uuid::nil(), Category::Request, 1, 70);
        assert_eq!(r.resolution, Some(Resolution::Unresolved));
        assert_eq!(r.tier, 0);
        let w = QueueEntry::new(Uuid::nil(), Category::Working, 2, 10);
        assert_eq!(w.resolution, None);
        assert_eq!(w.tier, 3);
        assert_eq!(w.state, EntryState::New);
    }

    #[test]
    fn round_trip_and_optional_fields() {
        let e = QueueEntry::new(Uuid::nil(), Category::Error, 5, 80);
        let s = serde_json::to_string(&e).unwrap();
        assert!(!s.contains("resolution"));
        let back: QueueEntry = serde_json::from_str(&s).unwrap();
        assert_eq!(e, back);

        let minimal = json!({
            "event_id": Uuid::nil(), "tier": 3, "score": 10, "state": "new", "seq": 1
        });
        let back: QueueEntry = serde_json::from_value(minimal).unwrap();
        assert!(!back.superseded && back.escalation_level == 0 && back.resolution.is_none());
    }

    #[test]
    fn wire_names() {
        assert_eq!(
            serde_json::to_value(EntryState::Dismissed).unwrap(),
            json!("dismissed")
        );
        assert_eq!(
            serde_json::to_value(Resolution::Approved).unwrap(),
            json!("approved")
        );
    }

    #[test]
    fn liveness_and_ordering() {
        let mut a = QueueEntry::new(Uuid::nil(), Category::Working, 1, 90);
        let b = QueueEntry::new(Uuid::nil(), Category::Completed, 2, 5);
        let c = QueueEntry::new(Uuid::nil(), Category::Working, 3, 90);
        // Lower tier wins regardless of score.
        assert!(b.order_key() < a.order_key());
        // Same tier and score: newer seq first.
        assert!(c.order_key() < a.order_key());

        assert!(a.is_live());
        a.state = EntryState::Dismissed;
        assert!(!a.is_live());
        let mut d = QueueEntry::new(Uuid::nil(), Category::Working, 4, 1);
        d.superseded = true;
        assert!(!d.is_live());
    }
}
